apis/apps/v1alpha1/opsrequest_webhook.go

```diff
@@ -416,12 +416,134 @@ func (r *OpsRequest) validateHorizontalScaling(_ context.Context, _ client.Clien
+		}
+		if !isSharding && replicaChanger.ReplicaChanges != nil && len(offlineOrOnlineInsNames) > int(*replicaChanger.ReplicaChanges) {
+			return fmt.Errorf(`the length of %s can't be greater than the "replicaChanges" for the component`, hScaleInstanceFieldName)
+		}
+		offlineOrOnlineInsCountMap := r.CountOfflineOrOnlineInstances(clusterName, hScale.ComponentName, offlineOrOnlineInsNames)
+		insTplChangeMap := map[string]int32{}
+		allReplicaChanges := int32(0)
+		for _, v := range replicaChanger.Instances {
+			compInsReplicas, ok := compInsTplMap[v.Name]
+			if !ok {
+				return fmt.Errorf(`%s cannot find the instance template "%s" in component "%s"`,
+					msgPrefix, v.Name, hScale.ComponentName)
+			}
+			if isScaleIn && v.ReplicaChanges > compInsReplicas {
+				return fmt.Errorf(`%s "replicaChanges" of instanceTemplate "%s" can't be greater than %d`,
+					msgPrefix, v.Name, compInsReplicas)
+			}
+			allReplicaChanges += v.ReplicaChanges
+			insTplChangeMap[v.Name] = v.ReplicaChanges
+		}
+		for insTplName, replicaCount := range offlineOrOnlineInsCountMap {
+			replicaChanges, ok := insTplChangeMap[insTplName]
+			if !ok {
+				allReplicaChanges += replicaCount
+				continue
+			}
+			if replicaChanges < replicaCount {
+				return fmt.Errorf(`"replicaChanges" can't be less than %d when %d instances of the instance template "%s" are configured in %s`,
+					replicaCount, replicaCount, insTplName, hScaleInstanceFieldName)
+			}
+		}
+		for _, insTpl := range newInstances {
+			if _, ok := compInsTplMap[insTpl.Name]; ok {
+				return fmt.Errorf(`new instance template "%s" already exists in component "%s"`, insTpl.Name, hScale.ComponentName)
+			}
+			allReplicaChanges += insTpl.GetReplicas()
+		}
+		if !isSharding && replicaChanger.ReplicaChanges != nil && allReplicaChanges > *replicaChanger.ReplicaChanges {
+			return fmt.Errorf(`%s "replicaChanges" can't be less than the sum of "replicaChanges" for specified instance templates`, msgPrefix)
+		}
+		return nil
+	}
+	if scaleIn != nil {
+		if err := validateHScaleOperation(scaleIn.ReplicaChanger, nil, scaleIn.OnlineInstancesToOffline, true); err != nil {
+			return err
+		}
+		if scaleIn.ReplicaChanges != nil && *scaleIn.ReplicaChanges > compSpec.Replicas {
+			return fmt.Errorf(`"scaleIn.replicaChanges" can't be greater than %d for component "%s"`, compSpec.Replicas, hScale.ComponentName)
+		}
+	}
+	if scaleOut != nil {
+		if err := validateHScaleOperation(scaleOut.ReplicaChanger, scaleOut.NewInstances, scaleOut.OfflineInstancesToOnline, false); err != nil {
+			return err
+		}
+		if len(scaleOut.OfflineInstancesToOnline) > 0 {
+			offlineInstanceSet := sets.New(compSpec.OfflineInstances...)
+			for _, offlineInsName := range scaleOut.OfflineInstancesToOnline {
+				if _, ok := offlineInstanceSet[offlineInsName]; !ok {
+					return fmt.Errorf(`cannot find the offline instance "%s" in component "%s" for scaleOut operation`, offlineInsName, hScale.ComponentName)
+				}
+			}
+		}
+	}
```

Comment:
should add one more check, if scaleOut.OfflineInstancesToOnline and scaleIn.OnlineInstancesToOfflineInstances have intersection, should be invalid.

Reply:
OfflineInstancesToOnline must be in the componentSpec.offlineInstances, and onlineInstancesToOffline is must be a onlineInstances, so they won't have any intersection.